sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DisposableMemberInNonDisposableClass.cs

```diff
@@ -28,102 +27,86 @@
-                return methodNodes.
-                    OfType<AssignmentExpressionSyntax>().
-                    Where(n => n.IsKind(SyntaxKind.SimpleAssignmentExpression) && n.Right is ObjectCreationExpressionSyntax).
-                    Select(n => semanticModel.GetSymbolInfo(n.Left).Symbol).
-                    OfType<IFieldSymbol>();
-            }
-            return empty;
+            var disposableFields = namedType.GetMembers()
+                                            .OfType<IFieldSymbol>()
+                                            .Where(fs => fs.IsNonStaticNonPublicDisposableField(analysisContext.Compilation.GetLanguageVersion())).ToHashSet();
+
+            var disposableFieldsWithInitializer = disposableFields.Where(IsOwnerSinceDeclaration);
+
+            var otherInitializationsOfFields = namedType.GetMembers()
+                                                        .OfType<IMethodSymbol>()
+                                                        .SelectMany(m => GetAssignmentsToFieldsIn(m, analysisContext.Compilation))
+                                                        .Where(f => disposableFields.Contains(f));
+
+            return string.Join(", ", disposableFieldsWithInitializer.Union(otherInitializationsOfFields)
+                                                                    .Distinct()
+                                                                    .Select(symbol => $"'{symbol.Name}'")
+                                                                    .OrderBy(name => name));
         }
 
-        private bool IsOwnerSinceDeclaration(IFieldSymbol f)
+        private static IEnumerable<IFieldSymbol> GetAssignmentsToFieldsIn(ISymbol m, Compilation compilation)
         {
-            var syntaxRef = f.DeclaringSyntaxReferences.SingleOrDefault();
-            if (syntaxRef == null)
+            if (m.DeclaringSyntaxReferences.Length != 1
+                || !(m.DeclaringSyntaxReferences[0].GetSyntax() is BaseMethodDeclarationSyntax method)
+                || !method.HasBodyOrExpressionBody())
             {
-                return false;
+                return Enumerable.Empty<IFieldSymbol>();
             }
-            if (syntaxRef.GetSyntax() is VariableDeclaratorSyntax varDeclarator)
-            {
-                return varDeclarator.Initializer != null && varDeclarator.Initializer.Value is ObjectCreationExpressionSyntax;
-            }
-            return false;
+
+            return method.GetBodyDescendantNodes()
+                         .OfType<AssignmentExpressionSyntax>()
+                         .Where(n => n.IsKind(SyntaxKind.SimpleAssignmentExpression) && n.Right is ObjectCreationExpressionSyntax)
+                         .Select(n => compilation.GetSemanticModel(method.SyntaxTree).GetSymbolInfo(n.Left).Symbol)
+                         .OfType<IFieldSymbol>();
         }
 
+        private static bool IsOwnerSinceDeclaration(ISymbol symbol) =>
```

Comment:
nitpick: IsOwnerSinceDeclaration should be above GetAssignmentsToFieldsIn because it's called first

Reply:
I would avoid this kind of conventions. Code changes, often methods are called from multiple places and the order in which the methods are called should not define the order in which they are defined.